### PR TITLE
Use clang-format/clang-tidy version 17.0.6

### PR DESCRIPTION
**Pull request by BruceForstall**
These are now uploaded in a new location in JIT Azure Storage for three platforms currently:
- windows-x64
- linux-x64
- osx-arm64

So jit-format will work on any of these platforms now.

Note that osx-x64 is no longer supported for clang-format/clang-tidy.

This includes https://github.com/dotnet/jitutils/pull/407.

### CODE REVIEW

**Comment by BruceForstall** [comment]
This requires https://github.com/dotnet/runtime/pull/100498 (which also depends on this change).

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib

File: src/jit-format/jit-format.cs [hunk 1]
```diff
@@ -149,9 +150,41 @@ private void validate()
...
+                    }
+
+                    if (_verbose)
+                    {
+                        Console.WriteLine("Process architecture is {0}", _arch);
+                    }
+                }
+                else
+                {
+                    _arch = _arch.ToLower();
```
Comment (am11) [inline src/jit-format/jit-format.cs:187]:
```suggestion
                    _arch = _arch.ToLowerInvariant();
```

Comment (am11) [inline src/jit-format/jit-format.cs:187]:
```suggestion
                    _arch = _arch.ToLowerInvariant();
```

File: src/jit-format/jit-format.cs [hunk 2]
```diff
@@ -178,8 +211,12 @@ private void validate()
                         Console.WriteLine("Operating system is {0}", _os);
                     }
                 }
+                else
+                {
+                    _os = _os.ToLower();
```
Comment (am11) [inline src/jit-format/jit-format.cs:216]:
```suggestion
                    _os = _os.ToLowerInvariant();
```

File: bootstrap.sh [hunk 3]
```diff
@@ -70,28 +70,24 @@ function download_tools {
...
+    # or:
+    #      RID:         linux-x64
+    #
+    # If the RID doesn't work directly, it could be used to pick the appropriate platform.
+
+    clangToolsRootUrl="https://clrjit2.blob.core.windows.net/clang-tools"
+    clangVersion="17.0.6"
+
+    clangPlatform=$(dotnet --info |grep RID:)
+    clangPlatform=${clangPlatform##*RID:* }
```
Comment (am11) [inline bootstrap.sh:85]:
```suggestion
    clangPlatform="$(dotnet --info | grep 'RID:')"
    clangPlatform="${clangPlatform##*RID:*}"
```

File: bootstrap.cmd [hunk 4]
```diff
@@ -131,24 +131,29 @@ IF %errorlevel% NEQ 0 GOTO DownloadTools
 
 REM We found the tools on the path; now make sure the versions are good.
 
-clang-format --version | findstr 3.8 > NUL
-If %errorlevel% EQU 0 GOTO build_done
+clang-format --version | findstr /c:"version 17.0.6" > NUL
```
Comment (am11) [inline bootstrap.cmd:134]: nit: we can define `_clang_version` earlier in the script and use the variable:

```suggestion
set _clang_version=17.0.6
clang-format --version | findstr /c:"version %_clang_version%" > NUL
```

File: bootstrap.cmd [hunk 5]
```diff
@@ -131,24 +131,29 @@ IF %errorlevel% NEQ 0 GOTO DownloadTools
...
+IF %errorlevel% EQU 0 GOTO build_done
 
-echo jit-format requires clang-format and clang-tidy version 3.8.*. Currently installed:
+echo jit-format requires clang-format and clang-tidy version 17.0.6. Currently installed:
```
Comment (am11) [inline bootstrap.cmd:137]:
```suggestion
echo jit-format requires clang-format and clang-tidy version %_clang_version%. Currently installed:
```

File: bootstrap.cmd [hunk 6]
```diff
@@ -131,24 +131,29 @@ IF %errorlevel% NEQ 0 GOTO DownloadTools
...
 clang-format --version
 clang-tidy --version
-echo Please install version 3.8.* and put the tools on the PATH to use jit-format.
-echo Tools can be found at https://llvm.org/releases/download.html#3.8.0
+echo Please install version 17.0.6 and put the tools on the PATH to use jit-format.
+echo Tools can be found at https://github.com/llvm/llvm-project/releases/tag/llvmorg-17.0.6
```
Comment (am11) [inline bootstrap.cmd:141]:
```suggestion
echo Please install version %_clang_version% and put the tools on the PATH to use jit-format.
echo Tools can be found at https://github.com/llvm/llvm-project/releases/tag/llvmorg-%_clang_version%
```

File: bootstrap.cmd [hunk 7]
```diff
@@ -131,24 +131,29 @@ IF %errorlevel% NEQ 0 GOTO DownloadTools
...
 
 :DownloadTools
 
 :: Download clang-format and clang-tidy
 echo Downloading formatting tools
 
-call :download_url clang-format "https://clrjit.blob.core.windows.net/clang-tools/windows/clang-format.exe" bin\clang-format.exe
+set _azdo_root=https://clrjit2.blob.core.windows.net/clang-tools
+set _clang_version=17.0.6
```
Comment (am11) [inline bootstrap.cmd:149]:
```suggestion
```

File: bootstrap.sh [hunk 8]
```diff
@@ -119,7 +114,7 @@ function download_tools {
 
     if [ ! -f bin/clang-format -o ! -f bin/clang-tidy ]; then
         echo "Either clang-tidy or clang-format was not installed. Please install and put them on the PATH to use jit-format."
-        echo "Tools can be found at https://llvm.org/releases/download.html#3.8.0"
+        echo "Tools can be found at https://github.com/llvm/llvm-project/releases/tag/llvmorg-17.0.6"
```
Comment (am11) [inline bootstrap.sh:117]:
```suggestion
        echo "Tools can be found at https://github.com/llvm/llvm-project/releases/tag/llvmorg-$clangVersion"
```

File: bootstrap.sh [hunk 9]
```diff
@@ -204,12 +199,12 @@ if ! hash clang-format 2>/dev/null || ! hash clang-tidy 2>/dev/null; then
     download_tools
     exit_code=$?
 else
-    if ! clang-format --version | grep -q 3.8 || ! clang-tidy --version | grep -q 3.8; then
-        echo "jit-format requires clang-format and clang-tidy version 3.8.*. Currently installed: "
+    if ! clang-format --version | grep -q "version 17.0.6" || ! clang-tidy --version | grep -q "version 17.0.6"; then
+        echo "jit-format requires clang-format and clang-tidy version 17.0.6. Currently installed: "
```
Comment (am11) [inline bootstrap.sh:203]:
```suggestion
    if ! clang-format --version | grep -q "version $clangVersion" || ! clang-tidy --version | grep -q "version $clangVersion"; then
        echo "jit-format requires clang-format and clang-tidy version $clangVersion. Currently installed: "
```

File: bootstrap.sh [hunk 10]
```diff
@@ -204,12 +199,12 @@ if ! hash clang-format 2>/dev/null || ! hash clang-tidy 2>/dev/null; then
...
         clang-format --version
         clang-tidy --version
 
-        echo "Installing version 3.8 of clang tools"
+        echo "Installing clang tools"
```
Comment (am11) [inline bootstrap.sh:207]:
```suggestion
        echo "Installing clang $clangVersion tools"
```

**Comment by BruceForstall** [comment]
> FWIW, the bleeding edge version is https://github.com/llvm/llvm-project/releases/tag/llvmorg-18.1.2, in case we want to opt for latest stable. Apple Clang has different versioning scheme, which does not correspond to llvmorg versions; current latest satble version on fully updated osx-arm64 machine is 1500.3.9.4 which corresponds to llvm 16.0.0 (https://en.wikipedia.org/wiki/Xcode#Xcode_15.0_-_(since_visionOS_support)_2).

@am11 Thanks for the review!

I went for 17.0.6 instead of 18 partially because there was prebuilt Ubuntu 22.04 binaries. And it's new but not too "bleeding edge". As for Apple, well, the LLVM github release does have osx-arm64 built bits.

**Review by AndyAyersMS — APPROVED** [review]

**Pull request merged**